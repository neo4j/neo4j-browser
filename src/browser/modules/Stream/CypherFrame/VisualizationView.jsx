/*
 * Copyright (c) 2002-2020 "Neo4j,"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import neo4j from 'neo4j-driver'
import React, { Component } from 'react'
import { connect } from 'react-redux'
import { deepEquals } from 'services/utils'
import * as grassActions from 'shared/modules/grass/grassDuck'
import bolt from 'services/bolt/bolt'
import { withBus } from 'react-suber'
import { ExplorerComponent } from '../../D3Visualization/components/Explorer'
import { StyledVisContainer } from './VisualizationView.styled'

import { CYPHER_REQUEST } from 'shared/modules/cypher/cypherDuck'
import { NEO4J_BROWSER_USER_ACTION_QUERY } from 'services/bolt/txMetadata'
import { getMaxFieldItems } from 'shared/modules/settings/settingsDuck'
import { resultHasTruncatedFields } from 'browser/modules/Stream/CypherFrame/helpers'

export class Visualization extends Component {
  state = {
    nodes: [],
    relationships: []
  }

  componentDidMount() {
    const { records = [] } = this.props.result
    if (records && records.length > 0) {
      this.populateDataToStateFromProps(this.props)
    }
  }

  shouldComponentUpdate(props, state) {
    return (
      this.props.updated !== props.updated ||
      !deepEquals(props.graphStyleData, this.props.graphStyleData) ||
      this.state.updated !== state.updated ||
      this.props.frameHeight !== props.frameHeight ||
      this.props.autoComplete !== props.autoComplete
    )
  }

  componentDidUpdate(prevProps) {
    if (
      this.props.updated !== prevProps.updated ||
      this.props.autoComplete !== prevProps.autoComplete
    ) {
      this.populateDataToStateFromProps(this.props)
    }
  }

  populateDataToStateFromProps(props) {
    const {
      nodes,
      relationships
    } = bolt.extractNodesAndRelationshipsFromRecordsForOldVis(
      props.result.records,
      true,
      props.maxFieldItems
    )
    const hasTruncatedFields = resultHasTruncatedFields(
      props.result,
      props.maxFieldItems
    )
    this.setState({
      nodes,
      relationships,
      hasTruncatedFields,
      updated: new Date().getTime()
    })
  }

  autoCompleteRelationships(existingNodes, newNodes) {
    if (this.props.autoComplete) {
      const existingNodeIds = existingNodes.map(node => parseInt(node.id))
      const newNodeIds = newNodes.map(node => parseInt(node.id))

      this.getInternalRelationships(existingNodeIds, newNodeIds)
        .then(graph => {
          this.autoCompleteCallback &&
            this.autoCompleteCallback(graph.relationships)
        })
        .catch(e => {})
    } else {
      this.autoCompleteCallback && this.autoCompleteCallback([])
    }
  }

  getNeighbours(id, currentNeighbourIds = []) {
    const query = `MATCH path = (a)--(o)
                   WHERE id(a) = ${id}
                   AND NOT (id(o) IN[${currentNeighbourIds.join(',')}])
                   RETURN path, size((a)--()) as c
                   ORDER BY id(o)
                   LIMIT ${this.props.maxNeighbours -
                     currentNeighbourIds.length}`
    return new Promise((resolve, reject) => {
      this.props.bus &&
        this.props.bus.self(
          CYPHER_REQUEST,
          { query: query, queryType: NEO4J_BROWSER_USER_ACTION_QUERY },
          response => {
            if (!response.success) {
              reject(new Error())
            } else {
              const count =
                response.result.records.length > 0
                  ? parseInt(response.result.records[0].get('c').toString())
                  : 0
              const resultGraph = bolt.extractNodesAndRelationshipsFromRecordsForOldVis(
                response.result.records,
                false,
                this.props.maxFieldItems
              )
              this.autoCompleteRelationships(
                this.graph._nodes,
                resultGraph.nodes
              )
              resolve({ ...resultGraph, count: count })
            }
          }
        )
    })
  }

  getInternalRelationships(existingNodeIds, newNodeIds) {
    newNodeIds = newNodeIds.map(neo4j.int)
    existingNodeIds = existingNodeIds.map(neo4j.int)
    existingNodeIds = existingNodeIds.concat(newNodeIds)
    const query =
      'MATCH (a)-[r]->(b) WHERE id(a) IN $existingNodeIds AND id(b) IN $newNodeIds RETURN r;'
    return new Promise((resolve, reject) => {
      this.props.bus &&
        this.props.bus.self(
          CYPHER_REQUEST,
          {
            query,
            params: { existingNodeIds, newNodeIds },
            queryType: NEO4J_BROWSER_USER_ACTION_QUERY
          },
          response => {
            if (!response.success) {
              reject(new Error())
            } else {
              resolve({
                ...bolt.extractNodesAndRelationshipsFromRecordsForOldVis(
                  response.result.records,
                  false,
                  this.props.maxFieldItems
                )
              })
            }
          }
        )
    })
  }

  setGraph(graph) {
    this.graph = graph
    this.autoCompleteRelationships([], this.graph._nodes)
  }

  render() {
    if (!this.state.nodes.length) return null

    return (
      <StyledVisContainer fullscreen={this.props.fullscreen}>
        <ExplorerComponent
          maxNeighbours={this.props.maxNeighbours}
          hasTruncatedFields={this.state.hasTruncatedFields}
          initialNodeDisplay={this.props.initialNodeDisplay}
          graphStyleData={this.props.graphStyleData}
          updateStyle={this.props.updateStyle}
          getNeighbours={this.getNeighbours.bind(this)}
          nodes={this.state.nodes}
          relationships={this.state.relationships}
          fullscreen={this.props.fullscreen}
          frameHeight={this.props.frameHeight}
          assignVisElement={this.props.assignVisElement}
          getAutoCompleteCallback={callback => {
            this.autoCompleteCallback = callback
          }}
          setGraph={this.setGraph.bind(this)}
        />
      </StyledVisContainer>
    )
  }
}

const mapStateToProps = state => {
  return {
    graphStyleData: grassActions.getGraphStyleData(state),
    maxFieldItems: getMaxFieldItems(state)
  }
}

const mapDispatchToProps = dispatch => {
  return {
    updateStyle: graphStyleData => {
      dispatch(grassActions.updateGraphStyleData(graphStyleData))
    }
  }
}

export const VisualizationConnectedBus = withBus(
  connect(mapStateToProps, mapDispatchToProps)(Visualization)
)
