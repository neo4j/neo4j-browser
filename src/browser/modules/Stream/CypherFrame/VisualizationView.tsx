/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import neo4j from 'neo4j-driver'
import React, { Component } from 'react'
import { connect } from 'react-redux'
import { withBus } from 'react-suber'
import { Action, Dispatch } from 'redux'
import { Bus } from 'suber'

import { GraphModel } from 'graph-visualization'

import Explorer from '../../D3Visualization/components/Explorer'
import { StyledVisContainer } from './VisualizationView.styled'
import { resultHasTruncatedFields } from 'browser/modules/Stream/CypherFrame/helpers'
import bolt from 'services/bolt/bolt'
import {
  BasicNode,
  BasicNodesAndRels,
  BasicRelationship
} from 'services/bolt/boltMappings'
import { NEO4J_BROWSER_USER_ACTION_QUERY } from 'services/bolt/txMetadata'
import { deepEquals } from 'services/utils'
import { GlobalState } from 'shared/globalState'
import { CYPHER_REQUEST } from 'shared/modules/cypher/cypherDuck'
import * as grassActions from 'shared/modules/grass/grassDuck'
import { getMaxFieldItems } from 'shared/modules/settings/settingsDuck'

type VisualizationState = {
  updated: number
  nodes: BasicNode[]
  relationships: BasicRelationship[]
  hasTruncatedFields: boolean
}

export type VisualizationProps = {
  result: any
  graphStyleData: any
  updated: number
  autoComplete: boolean
  maxNeighbours: number
  bus: Bus
  maxFieldItems: number
  initialNodeDisplay: number
  isFullscreen: boolean
  updateStyle: (style: any) => void
  assignVisElement: (v: any) => void
}

export class Visualization extends Component<
  VisualizationProps,
  VisualizationState
> {
  autoCompleteCallback: ((rels: BasicRelationship[]) => void) | undefined
  graph: GraphModel | undefined
  state: VisualizationState = {
    nodes: [],
    relationships: [],
    updated: 0,
    hasTruncatedFields: false
  }

  componentDidMount(): void {
    const { records = [] } = this.props.result
    if (records && records.length > 0) {
      this.populateDataToStateFromProps(this.props)
    }
  }

  shouldComponentUpdate(
    props: VisualizationProps,
    state: VisualizationState
  ): boolean {
    return (
      this.props.updated !== props.updated ||
      this.props.isFullscreen !== props.isFullscreen ||
      !deepEquals(props.graphStyleData, this.props.graphStyleData) ||
      this.state.updated !== state.updated ||
      this.props.autoComplete !== props.autoComplete
    )
  }

  componentDidUpdate(prevProps: VisualizationProps): void {
    if (
      this.props.updated !== prevProps.updated ||
      this.props.autoComplete !== prevProps.autoComplete
    ) {
      this.populateDataToStateFromProps(this.props)
    }
  }

  populateDataToStateFromProps(props: VisualizationProps): void {
    const { nodes, relationships } =
      bolt.extractNodesAndRelationshipsFromRecordsForOldVis(
        props.result.records,
        true,
        props.maxFieldItems
      )
    const hasTruncatedFields = resultHasTruncatedFields(
      props.result,
      props.maxFieldItems
    )
    this.setState({
      nodes,
      relationships,
      hasTruncatedFields,
      updated: new Date().getTime()
    })
  }

  autoCompleteRelationships(
    existingNodes: { id: string }[],
    newNodes: { id: string }[]
  ): void {
    if (this.props.autoComplete) {
      const existingNodeIds = existingNodes.map(node => parseInt(node.id))
      const newNodeIds = newNodes.map(node => parseInt(node.id))

      this.getInternalRelationships(existingNodeIds, newNodeIds)
        .then(graph => {
          this.autoCompleteCallback &&
            this.autoCompleteCallback(graph.relationships)
        })
        .catch(() => undefined)
    } else {
      this.autoCompleteCallback && this.autoCompleteCallback([])
    }
  }

  getNeighbours(
    id: string,
    currentNeighbourIds: string[] = []
  ): Promise<BasicNodesAndRels & { allNeighboursCount: number }> {
    const maxNewNeighbours =
      this.props.maxNeighbours - currentNeighbourIds.length

    const query =
      maxNewNeighbours > 0
        ? `MATCH (a) WHERE id(a) = ${id}
WITH a, size([(a)--() | 1]) AS allNeighboursCount
MATCH path = (a)--(o) WHERE NOT id(o) IN [${currentNeighbourIds.join(',')}]
RETURN path, allNeighboursCount
ORDER BY id(o)
LIMIT ${maxNewNeighbours}`
        : `MATCH p=(a)--() WHERE id(a) = ${id} RETURN count(p) as allNeighboursCount`

    return new Promise((resolve, reject) => {
      this.props.bus &&
        this.props.bus.self(
          CYPHER_REQUEST,
          { query: query, queryType: NEO4J_BROWSER_USER_ACTION_QUERY },
          (response: any) => {
            if (!response.success) {
              reject(new Error())
            } else {
              const allNeighboursCount =
                response.result.records.length > 0
                  ? parseInt(
                      response.result.records[0]
                        .get('allNeighboursCount')
                        .toString()
                    )
                  : 0
              const resultGraph =
                bolt.extractNodesAndRelationshipsFromRecordsForOldVis(
                  response.result.records,
                  false,
                  this.props.maxFieldItems
                )
              this.autoCompleteRelationships(
                this.graph?.nodes() || [],
                resultGraph.nodes
              )
              resolve({ ...resultGraph, allNeighboursCount })
            }
          }
        )
    })
  }

  getInternalRelationships(
    rawExistingNodeIds: number[],
    rawNewNodeIds: number[]
  ): Promise<BasicNodesAndRels> {
    const newNodeIds = rawNewNodeIds.map(neo4j.int)
    const existingNodeIds = rawExistingNodeIds.map(neo4j.int).concat(newNodeIds)
    const query =
      'MATCH (a)-[r]->(b) WHERE id(a) IN $existingNodeIds AND id(b) IN $newNodeIds RETURN r;'
    return new Promise((resolve, reject) => {
      this.props.bus &&
        this.props.bus.self(
          CYPHER_REQUEST,
          {
            query,
            params: { existingNodeIds, newNodeIds },
            queryType: NEO4J_BROWSER_USER_ACTION_QUERY
          },
          (response: any) => {
            if (!response.success) {
              reject(new Error())
            } else {
              resolve({
                ...bolt.extractNodesAndRelationshipsFromRecordsForOldVis(
                  response.result.records,
                  false,
                  this.props.maxFieldItems
                )
              })
            }
          }
        )
    })
  }

  setGraph(graph: GraphModel): void {
    this.graph = graph
    this.autoCompleteRelationships([], this.graph.nodes())
  }

  render(): React.ReactNode {
    if (!this.state.nodes.length) return null

    return (
      <StyledVisContainer isFullscreen={this.props.isFullscreen}>
        <Explorer
          maxNeighbours={this.props.maxNeighbours}
          hasTruncatedFields={this.state.hasTruncatedFields}
          initialNodeDisplay={this.props.initialNodeDisplay}
          graphStyleData={this.props.graphStyleData}
          updateStyle={this.props.updateStyle}
          getNeighbours={this.getNeighbours.bind(this)}
          nodes={this.state.nodes}
          relationships={this.state.relationships}
          isFullscreen={this.props.isFullscreen}
          assignVisElement={this.props.assignVisElement}
          getAutoCompleteCallback={(
            callback: (rels: BasicRelationship[]) => void
          ) => {
            this.autoCompleteCallback = callback
          }}
          setGraph={this.setGraph.bind(this)}
        />
      </StyledVisContainer>
    )
  }
}

const mapStateToProps = (state: GlobalState) => ({
  graphStyleData: grassActions.getGraphStyleData(state),
  maxFieldItems: getMaxFieldItems(state)
})

const mapDispatchToProps = (dispatch: Dispatch<Action>) => ({
  updateStyle: (graphStyleData: any) => {
    dispatch(grassActions.updateGraphStyleData(graphStyleData))
  }
})

export const VisualizationConnectedBus = withBus(
  connect(mapStateToProps, mapDispatchToProps)(Visualization)
)
