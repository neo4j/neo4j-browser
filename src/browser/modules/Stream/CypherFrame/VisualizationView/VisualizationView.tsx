/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import neo4j from 'neo4j-driver'
import React, { Component } from 'react'
import { connect } from 'react-redux'
import { withBus } from 'react-suber'
import { Action, Dispatch } from 'redux'
import { Bus } from 'suber'

import { GraphModel, GraphVisualizer } from 'neo4j-arc/graph-visualization'

import { StyledVisContainer } from './VisualizationView.styled'
import { resultHasTruncatedFields } from 'browser/modules/Stream/CypherFrame/helpers'
import bolt from 'services/bolt/bolt'
import {
  BasicNode,
  BasicNodesAndRels,
  BasicRelationship
} from 'neo4j-arc/common'
import { NEO4J_BROWSER_USER_ACTION_QUERY } from 'services/bolt/txMetadata'
import { deepEquals } from 'neo4j-arc/common'
import { GlobalState } from 'shared/globalState'
import { CYPHER_REQUEST } from 'shared/modules/cypher/cypherDuck'
import * as grassActions from 'shared/modules/grass/grassDuck'
import {
  getMaxFieldItems,
  shouldShowWheelZoomInfo,
  update as updateSettings
} from 'shared/modules/settings/settingsDuck'
import {
  getNodePropertiesExpandedByDefault,
  setNodePropertiesExpandedByDefault
} from 'shared/modules/frames/framesDuck'
import { DetailsPane } from './PropertiesPanelContent/DetailsPane'
import OverviewPane from './PropertiesPanelContent/OverviewPane'

type VisualizationState = {
  updated: number
  nodes: BasicNode[]
  relationships: BasicRelationship[]
  hasTruncatedFields: boolean
  nodeLimitHit: boolean
}

export type VisualizationProps = {
  result: any
  graphStyleData: any
  updated: number
  autoComplete: boolean
  maxNeighbours: number
  bus: Bus
  maxFieldItems: number
  initialNodeDisplay: number
  isFullscreen: boolean
  updateStyle: (style: any) => void
  assignVisElement: (v: any) => void
  nodePropertiesExpandedByDefault: boolean
  setNodePropertiesExpandedByDefault: (expandedByDefault: boolean) => void
  wheelZoomInfoMessageEnabled: boolean
  disableWheelZoomInfoMessage: () => void
}

export class Visualization extends Component<
  VisualizationProps,
  VisualizationState
> {
  autoCompleteCallback: ((rels: BasicRelationship[]) => void) | undefined
  graph: GraphModel | undefined
  state: VisualizationState = {
    nodes: [],
    relationships: [],
    updated: 0,
    nodeLimitHit: false,
    hasTruncatedFields: false
  }

  componentDidMount(): void {
    const { records = [] } = this.props.result
    if (records && records.length > 0) {
      this.populateDataToStateFromProps(this.props)
    }
  }

  shouldComponentUpdate(
    props: VisualizationProps,
    state: VisualizationState
  ): boolean {
    return (
      this.props.updated !== props.updated ||
      this.props.isFullscreen !== props.isFullscreen ||
      !deepEquals(props.graphStyleData, this.props.graphStyleData) ||
      this.state.updated !== state.updated ||
      this.props.autoComplete !== props.autoComplete ||
      this.props.wheelZoomInfoMessageEnabled !==
        props.wheelZoomInfoMessageEnabled
    )
  }

  componentDidUpdate(prevProps: VisualizationProps): void {
    if (
      this.props.updated !== prevProps.updated ||
      this.props.autoComplete !== prevProps.autoComplete
    ) {
      this.populateDataToStateFromProps(this.props)
    }
  }

  populateDataToStateFromProps(props: VisualizationProps): void {
    const { nodes, relationships } =
      bolt.extractNodesAndRelationshipsFromRecordsForOldVis(
        props.result.records,
        true,
        props.maxFieldItems
      )

    const { nodes: uniqNodes, nodeLimitHit } = deduplicateNodes(
      nodes,
      this.props.initialNodeDisplay
    )

    const uniqRels = nodeLimitHit
      ? relationships.filter(
          rel =>
            !!uniqNodes.find(node => node.id === rel.startNodeId) &&
            !!uniqNodes.find(node => node.id === rel.endNodeId)
        )
      : relationships

    const hasTruncatedFields = resultHasTruncatedFields(
      props.result,
      props.maxFieldItems
    )
    this.setState({
      nodes: uniqNodes,
      relationships: uniqRels,
      nodeLimitHit,
      hasTruncatedFields,
      updated: new Date().getTime()
    })
  }

  autoCompleteRelationships(
    existingNodes: { id: string }[],
    newNodes: { id: string }[]
  ): void {
    if (this.props.autoComplete) {
      const existingNodeIds = existingNodes.map(node => parseInt(node.id))
      const newNodeIds = newNodes.map(node => parseInt(node.id))

      this.getInternalRelationships(existingNodeIds, newNodeIds)
        .then(graph => {
          this.autoCompleteCallback &&
            this.autoCompleteCallback(graph.relationships)
        })
        .catch(() => undefined)
    } else {
      this.autoCompleteCallback && this.autoCompleteCallback([])
    }
  }

  getNeighbours(
    id: string,
    currentNeighbourIds: string[] = []
  ): Promise<BasicNodesAndRels & { allNeighboursCount: number }> {
    const maxNewNeighbours =
      this.props.maxNeighbours - currentNeighbourIds.length

    const query =
      maxNewNeighbours > 0
        ? `MATCH (a) WHERE id(a) = ${id}
WITH a, size([(a)--() | 1]) AS allNeighboursCount
MATCH path = (a)--(o) WHERE NOT id(o) IN [${currentNeighbourIds.join(',')}]
RETURN path, allNeighboursCount
ORDER BY id(o)
LIMIT ${maxNewNeighbours}`
        : `MATCH p=(a)--() WHERE id(a) = ${id} RETURN count(p) as allNeighboursCount`

    return new Promise((resolve, reject) => {
      this.props.bus &&
        this.props.bus.self(
          CYPHER_REQUEST,
          { query: query, queryType: NEO4J_BROWSER_USER_ACTION_QUERY },
          (response: any) => {
            if (!response.success) {
              reject(new Error())
            } else {
              const allNeighboursCount =
                response.result.records.length > 0
                  ? parseInt(
                      response.result.records[0]
                        .get('allNeighboursCount')
                        .toString()
                    )
                  : 0
              const resultGraph =
                bolt.extractNodesAndRelationshipsFromRecordsForOldVis(
                  response.result.records,
                  false,
                  this.props.maxFieldItems
                )
              this.autoCompleteRelationships(
                this.graph?.nodes() || [],
                resultGraph.nodes
              )
              resolve({ ...resultGraph, allNeighboursCount })
            }
          }
        )
    })
  }

  getInternalRelationships(
    rawExistingNodeIds: number[],
    rawNewNodeIds: number[]
  ): Promise<BasicNodesAndRels> {
    const newNodeIds = rawNewNodeIds.map(neo4j.int)
    const existingNodeIds = rawExistingNodeIds.map(neo4j.int).concat(newNodeIds)
    const query =
      'MATCH (a)-[r]->(b) WHERE id(a) IN $existingNodeIds AND id(b) IN $newNodeIds RETURN r;'
    return new Promise((resolve, reject) => {
      this.props.bus &&
        this.props.bus.self(
          CYPHER_REQUEST,
          {
            query,
            params: { existingNodeIds, newNodeIds },
            queryType: NEO4J_BROWSER_USER_ACTION_QUERY
          },
          (response: any) => {
            if (!response.success) {
              reject(new Error())
            } else {
              resolve({
                ...bolt.extractNodesAndRelationshipsFromRecordsForOldVis(
                  response.result.records,
                  false,
                  this.props.maxFieldItems
                )
              })
            }
          }
        )
    })
  }

  setGraph(graph: GraphModel): void {
    this.graph = graph
    this.autoCompleteRelationships([], this.graph.nodes())
  }

  render(): React.ReactNode {
    if (!this.state.nodes.length) return null

    return (
      <StyledVisContainer isFullscreen={this.props.isFullscreen}>
        <GraphVisualizer
          maxNeighbours={this.props.maxNeighbours}
          hasTruncatedFields={this.state.hasTruncatedFields}
          graphStyleData={this.props.graphStyleData}
          updateStyle={this.props.updateStyle}
          getNeighbours={this.getNeighbours.bind(this)}
          nodes={this.state.nodes}
          autocompleteRelationships={this.props.autoComplete ?? false}
          relationships={this.state.relationships}
          isFullscreen={this.props.isFullscreen}
          assignVisElement={this.props.assignVisElement}
          nodeLimitHit={this.state.nodeLimitHit}
          getAutoCompleteCallback={(
            callback: (rels: BasicRelationship[]) => void
          ) => {
            this.autoCompleteCallback = callback
          }}
          setGraph={this.setGraph.bind(this)}
          setNodePropertiesExpandedByDefault={
            this.props.setNodePropertiesExpandedByDefault
          }
          nodePropertiesExpandedByDefault={
            this.props.nodePropertiesExpandedByDefault
          }
          wheelZoomRequiresModKey={!this.props.isFullscreen}
          wheelZoomInfoMessageEnabled={
            this.props.wheelZoomInfoMessageEnabled && !this.props.isFullscreen
          }
          disableWheelZoomInfoMessage={this.props.disableWheelZoomInfoMessage}
          DetailsPaneOverride={DetailsPane}
          OverviewPaneOverride={OverviewPane}
          useGeneratedDefaultColors={false}
          initialZoomToFit
        />
      </StyledVisContainer>
    )
  }
}

const mapStateToProps = (state: GlobalState) => ({
  graphStyleData: grassActions.getGraphStyleData(state),
  maxFieldItems: getMaxFieldItems(state),
  nodePropertiesExpandedByDefault: getNodePropertiesExpandedByDefault(state),
  wheelZoomInfoMessageEnabled: shouldShowWheelZoomInfo(state)
})

const mapDispatchToProps = (dispatch: Dispatch<Action>) => ({
  disableWheelZoomInfoMessage: () => {
    dispatch(updateSettings({ showWheelZoomInfo: false }))
  },
  setNodePropertiesExpandedByDefault: (expandedByDefault: boolean) =>
    dispatch(setNodePropertiesExpandedByDefault(expandedByDefault)),
  updateStyle: (graphStyleData: any) => {
    dispatch(grassActions.updateGraphStyleData(graphStyleData))
  }
})

export const VisualizationConnectedBus = withBus(
  connect(mapStateToProps, mapDispatchToProps)(Visualization)
)

type DeduplicateHelper = {
  nodes: BasicNode[]
  taken: Record<string, boolean>
  nodeLimitHit: boolean
}

const deduplicateNodes = (
  nodes: BasicNode[],
  limit: number
): { nodes: BasicNode[]; nodeLimitHit: boolean } =>
  nodes.reduce(
    (all: DeduplicateHelper, curr: BasicNode) => {
      if (all.nodes.length === limit) {
        all.nodeLimitHit = true
      } else if (!all.taken[curr.id]) {
        all.nodes.push(curr)
        all.taken[curr.id] = true
      }
      return all
    },
    { nodes: [], taken: {}, nodeLimitHit: false }
  )
