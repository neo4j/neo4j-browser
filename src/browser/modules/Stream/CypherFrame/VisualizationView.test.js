/*
 * Copyright (c) 2002-2019 "Neo4j,"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import { render } from '@testing-library/react'
import neo4j from 'neo4j-driver'
import { Visualization } from './VisualizationView'

const mockEmptyResult = {
  records: []
}
const node = new neo4j.types.Node('1', ['Person'], {
  prop1: 'prop1'
})
const mockResult = {
  records: [{ keys: ['0'], __fields: [node], get: key => node }]
}

test('Visualization renders', () => {
  const { container } = render(<Visualization result={mockEmptyResult} />)
  expect(container).toMatchSnapshot()
})
test('Visualization renders with result', () => {
  const { container } = render(
    <Visualization updateStyle={() => {}} autoComplete result={mockResult} />
  )
  expect(container).toMatchSnapshot()
})
