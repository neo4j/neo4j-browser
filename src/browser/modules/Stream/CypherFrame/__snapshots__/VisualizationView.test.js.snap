// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visualization renders 1`] = `<div />`;

exports[`Visualization renders with result 1`] = `
<div>
  <div
    class="VisualizationViewstyled__StyledVisContainer-sc-1b47q51-0 dFbJTz"
  >
    <div
      class="one-legend-row styled__StyledFullSizeContainer-sc-13boxex-30 hrlJke"
      id="svg-vis"
    >
      <div
        class="one-row styled__StyledLegend-sc-13boxex-23 epTuur"
      >
        <div
          class="contracted styled__StyledLegendRow-sc-13boxex-22 btLVbI"
        >
          <ul
            class="styled__StyledInlineList-sc-13boxex-9 list-inline styled__StyledLegendInlineList-sc-13boxex-24 bXzLDX"
          >
            <li
              class="styled__StyledInlineListItem-sc-13boxex-10 styled__StyledLegendInlineListItem-sc-13boxex-25 bKGCvD"
              data-testid="viz-legend-labels"
            >
              <ul
                class="contents styled__StyledLegendContents-sc-13boxex-21 hvWlVe"
              >
                <li
                  class="styled__StyledInlineListItem-sc-13boxex-10 styled__StyledToken-sc-13boxex-15 token token-label styled__StyledLabelToken-sc-13boxex-16 jNYiQl"
                  style="background-color: rgb(247, 151, 103); color: rgb(255, 255, 255);"
                >
                  *
                  <span
                    class="count styled__StyledTokenCount-sc-13boxex-20 SIGMp"
                  >
                    (1)
                  </span>
                </li>
              </ul>
            </li>
            <li
              class="styled__StyledInlineListItem-sc-13boxex-10 styled__StyledLegendInlineListItem-sc-13boxex-25 bKGCvD"
              data-testid="viz-legend-labels"
            >
              <ul
                class="contents styled__StyledLegendContents-sc-13boxex-21 hvWlVe"
              >
                <li
                  class="styled__StyledInlineListItem-sc-13boxex-10 styled__StyledToken-sc-13boxex-15 token token-label styled__StyledLabelToken-sc-13boxex-16 jNYiQl"
                  style="background-color: rgb(201, 144, 192); color: rgb(255, 255, 255);"
                >
                  Person
                  <span
                    class="count styled__StyledTokenCount-sc-13boxex-20 SIGMp"
                  >
                    (1)
                  </span>
                </li>
              </ul>
            </li>
          </ul>
        </div>
      </div>
      <div
        class="styled__StyledSvgWrapper-sc-13boxex-0 emUVWV"
      >
        <svg
          class="neod3viz"
          viewBox="0 100 200 0"
        >
          <g
            transform="translate(0,0)"
          >
            <rect
              height="5000"
              style="fill: none; pointer-events: all;"
              transform="scale(1)"
              width="5000"
              x="-2500"
              y="-2500"
            />
            <g>
              <g
                class="layer relationships"
              />
              <g
                class="layer nodes"
              >
                <g
                  class="node"
                >
                  <circle
                    class="ring"
                    cx="0"
                    cy="0"
                    r="29"
                    stroke-width="8px"
                  />
                  <circle
                    class="outline"
                    cx="0"
                    cy="0"
                    fill="#C990C0"
                    r="25"
                    stroke="#b261a5"
                    stroke-width="2px"
                  />
                  <text
                    dy="1.5625"
                    fill="#FFFFFF"
                    font-family="streamline"
                    font-size="25"
                    pointer-events="none"
                    text-anchor="middle"
                  />
                  <text
                    fill="#FFFFFF"
                    font-size="10px"
                    pointer-events="none"
                    text-anchor="middle"
                    y="5"
                  >
                     proâ€¦
                  </text>
                </g>
              </g>
            </g>
          </g>
        </svg>
      </div>
      <div
        class="status-bar styled__StyledStatusBar-sc-13boxex-12 fBQkwD"
      >
        <div
          class="status styled__StyledStatus-sc-13boxex-13 lfPraq"
        >
          <div
            class="contracted inspector-footer styled__StyledInspectorFooter-sc-13boxex-5 fLefTL"
          >
            <ul
              class="inspector-footer-row styled__StyledInspectorFooterRow-sc-13boxex-6 cjMrGU"
              data-testid="vizInspector"
            >
              <ul
                class="list-inline styled__StyledInlineList-sc-13boxex-9 fWraRP"
              >
                <li
                  class="styled__StyledInlineListItem-sc-13boxex-10 pair styled__StyledInspectorFooterRowListPair-sc-13boxex-14 jeBqUF"
                >
                  <div
                    class="value styled__StyledInspectorFooterRowListValue-sc-13boxex-8 huEFTy"
                  >
                    Displaying 1 nodes, 0 relationships.
                  </div>
                </li>
              </ul>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
